Contributions from Taylor's Review of giving-tuesday-2022

jtfairbank opened the pull request:
Relates to #699

jtfairbank committed:
refactor: move fundraiser types from components/ to types/

jtfairbank committed:
refactor(fundraiser): use gatsby filesystem api to create fundraiser pages

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=840&src=https%3A%2F%2Fgithub.com%2Fdistributeaid%2Fdistributeaid.org">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=840&src=https%3A%2F%2Fgithub.com%2Fdistributeaid%2Fdistributeaid.org)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/distributeaid/distributeaid.org/840/2cfd5c37/b8e655c460ac9e19c7fff880ce8c05990affc3a1.svg)](https://app.codesee.io/r/reviews?pr=840&src=https%3A%2F%2Fgithub.com%2Fdistributeaid%2Fdistributeaid.org)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

jtfairbank committed:
refactor(gatsby): add resolver for gallery images (work in progress)

jtfairbank committed:
fix: select fundraiser gallery images from gatsby image sharp using a resolver.

jtfairbank committed:
refactor(gatsby): setup Photo node and set fundraiser.gallery to an array of Photos

coderbyheart commented on gatsby/fundraisers/transformers.ts:
```diff
@@ -55,8 +63,8 @@ export const createFundraiserSchemaCustomization = ({
-      gallery: [DAFundraiserPhoto]!,
-      allocations: [DAFundraiserAllocation]!
+      gallery: [DAPhoto!]!
+      allocations: [DAFundraiserAllocation!]!
```
```suggestion
      allocations: [DAFundraiserAllocation]!
```
Projects may have 0 allocations.

jtfairbank replied:
@coderbyheart I think that's ok?  Based on [what I read in the docs](https://graphql.org/learn/schema/#object-types-and-fields) the `[DAFundraiserAllocation!]!` means that that graphql will always return an array with 0 or more `DAFundraiserAllocation` entries inside of it.

So it can return `[]` or `[{allocation}, ...]`

but not `null` or `[{allocation}, null, ...]`

coderbyheart replied:
Ah, I forgot that it means that you can have `null` for `DAFundraiserAllocation` (without `!`)

jtfairbank committed:
feat: style the fundraiser pages

jtfairbank committed:
feat: style the fundraiser page WaysToDonate component (work in progress).

jtfairbank committed:
feat: finish styling the fundraiser page's WaysToDonate component.

jtfairbank committed:
Update from Forestry.io - Updated Forestry configuration

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/us-disaster-relief.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/ukraine-response.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/tech-for-good.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/sustainable-supply-chains.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/humanitarian-logistics.md
Taylor Fairbank updated content/blocks/fundraisers/european-refugee-relief.md

jtfairbank committed:
feat: add targets to fundraisers and a progress bar on the fundraiser page.

jtfairbank committed:
feat: add overall progress bar to the donate page.

jtfairbank committed:
fix: remove donate page header image

coderbyheart commented on .forestry/front_matter/templates/fundraiser.yml:
```diff
@@ -10,6 +10,17 @@ fields:
     min: 2
   label: Title
   description: The title of the fundraiser.
+- name: target
```
*comment:* I see targets are back. I think this is contra-productive providing targets without the explanation where this target comes from. None of the fundraisers have this explained in their texts. I also think this is not helpful for open-ended projects, which all of them are. In practice they can have unlimited funding needs. So if this is not tied to clear milestones I advise against providing targets.

jtfairbank replied:
@coderbyheart it's tricky.  After talking to Sara & Ellie they really wanted to set this up with donations going to each fundraiser, since they are organizing donors / aid matching around each fundraiser.

You're absolutely right that most if not all these fundraisers are ongoing, and have ongoing needs (or we'd know what to do with $XX,XXX more than the target).  For giving Tuesday, they want to focus on an achievable amount across all fundraisers for the 1-day push, so our initial targets will be for 1 month of expenses for each fundraiser.  However, we've repeatedly seen that framing things in terms of "what this covers for DA" doesn't do nearly as well as "what impact this creates".  So Ellie is going to come up with a key monthly stat per fundraiser to highlight the impact that achieving each target would create, along with a suggested donation amount to tie the individual donation to some portion of the overall impact ($20 lets us meet X need, etc).

In the future, I think it'd make sense to define "milestones" for each fundraiser.  This could describe what the funding is expected to go towards, paired with a description of the outcomes it'd lead to (X aid shipped, Y capability added, etc).  That'd also provide a nice history and make reporting easy.

What do you think about this approach?  Move forward with individual fundraiser links + targets for now, then build in milestones within / between fundraisers?

coderbyheart replied:
Yes, I think milestones are really the best way to document needs. This could be modeled similar to the `allocations`, so a project can have multiple of those.

jtfairbank committed:
Update from Forestry.io - Updated Forestry configuration

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/us-disaster-relief.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/ukraine-response.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/tech-for-good.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/sustainable-supply-chains.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/humanitarian-logistics.md

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/blocks/fundraisers/european-refugee-relief.md

jtfairbank committed:
feat: pre-calculate fundraiser allocation total during gatsby build phase

jtfairbank committed:
feat: add per-fundraiser donate links

jtfairbank committed:
style: the progress bar to work for the fundraiser cards & fundraiser page heder

jtfairbank committed:
chore: update the donate page's location in the /src/pages directory tree.

jtfairbank committed:
feat: use hover effects to highlight links on the donate page

jtfairbank committed:
fix: donate page e2e tests

jtfairbank committed:
style: the ways to donate section of the donate and fundraiser pages.

jtfairbank committed:
style: finishing touches on the donate page header.

jtfairbank committed:
fix: e2e donate tests and bank info modal closing

jtfairbank committed:
feat: feat: show a all the images instead of them one at a time on the fundraiser pages.

jtfairbank committed:
fix: progress bar content order

jtfairbank committed:
fix: ignore .stories.tsx files in subdirectories of /src/pages/

jtfairbank committed:
feat: mobile responsive styling for the updated donate and fundraising pages.

jtfairbank commented on src/stylesheets/donate.css:
```diff
@@ -6,183 +6,3 @@ main.donate {
   color: var(--color-da);
   background-color: white;
```
Hey @coderbyheart ended up converting most of the css to tailwind.  Is this last `main.donate` style needed?

jtfairbank commented:
Hey @coderbyheart this is ready for your review!  Bit of time pressure to get the initial donate pages launched so we can start lining up matching donations for Giving Tuesday + use the links when prepping marketing / social media.  So it'd be helpful to get two types of feedback: "need to change now" vs. "can launch but update this over the next week before Giving Tuesday".

jtfairbank committed:
Update from Forestry.io
Taylor Fairbank updated content/pages/donate.md

jtfairbank committed:
fix(style): fix max-width issue cause off-centering in the fundraiser page's title.

Commit:
Update from Forestry.io - Updated Forestry configuration

Commit:
Update from Forestry.io
Distribute Aid Admin updated content/pages/donate.md

coderbyheart commented:
There are a bunch of errors when the dev server is started:

```
warn Dropping content block type="block-image-with-caption", since it is not implemented yet.
warn Dropping empty section: #0
warn Dropping unknown content block: type="block-card"
warn Dropping unknown content block: type="block-card"
warn Dropping unknown content block: type="block-card"
warn Dropping empty section: #1

 ERROR 

Page title="Connecting Communities" is empty after processing.

```

coderbyheart commented:
e2e tests don't pass but I guess you want to fix that later.

coderbyheart commented:
Most likely caused by JS errors on campaign page:

![image](https://user-images.githubusercontent.com/188915/202819505-3e588213-585a-44d4-ae5a-7527a68d51e8.png)

jtfairbank commented:
@coderbyheart yeah I definitely want to do a testing pass on the donate / fundraising page code.  Probably after this week tho if it's stable enough to make it thru giving tuesday.

The gatsby build is logging a lot of content errors / warnings to let us know what to expect out of the final build.  I think many of these will go away once we implement other content blocks (image, youtube, etc).  We should def look into a better way to report content issues tho, would it be possible to log them during gatsby build and dump the log file to a slack channel or something?  Basically, if there is a content issue how can we let our content contributors know without a dev having to catch it in the build logs.

Made an issue for the bug you pointed out: https://github.com/distributeaid/distributeaid.org/issues/852#issue-1456742297

jtfairbank commented:
Really great working with you on this one!

coderbyheart commented:
> would it be possible to log them during gatsby build and dump the log file to a slack channel or something?

It would be possible but logging without actions is just noise and distraction. Nobody will be able to distinguish which ones are new and which ones can be ignored.
